File: framework/java/src/main/java/soda/unittest/validate/ValidatorFactory.java
package soda.unittest.validate;

import soda.unittest.Validators;
import soda.unittest.work.parse.TypeRef;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiPredicate;
import java.util.function.Supplier;

public class ValidatorFactory {

    private static final Map<Type, Supplier<Validator<?>>> factoryMap = new HashMap<>();

    private static void registerFactory(Type type, Supplier<Validator<?>> fact) {
        factoryMap.put(type, fact);
    }

    private static void registerFactory(TypeRef<?> typeRef, Supplier<Validator<?>> fact) {
        factoryMap.put(typeRef.getRefType(), fact);
    }

    private static <T> Validator<T> wrap(BiPredicate<T, T> b) {
        return b::test;
    }

    static {
        registerFactory(double[].class, () -> wrap(Validators.forDoubleArray(true)));
        registerFactory(new TypeRef<List<Double>>() {}, () -> wrap(Validators.forList(Double.class, true)));
        registerFactory(double[][].class, () -> wrap(Validators.forDoubleArray2d(true, true)));
        registerFactory(new TypeRef<List<List<Double>>>() {}, () -> wrap(Validators.forList2d(Double.class, true, true)));
    }

    @SuppressWarnings("unchecked")
    public static <T> Validator<T> create(Type type) {
        var fact = factoryMap.get(type);
        if (fact != null) {
            return (Validator<T>) fact.get();
        }
        return (a, b) -> FeatureFactory.create(type).isEqual(a, b);
    }

}

File: framework/scala/src/main/scala/soda/scala/unittest/ValidatorFactory.scala
package soda.scala.unittest

import scala.collection.mutable
import scala.reflect.runtime.universe._

object ValidatorFactory {

  private val factoryMap = mutable.Map[Type, () => (_, _) => Boolean]()

  private def registerFactory[T](factory: () => (T, T) => Boolean)(implicit tt: TypeTag[T]): Unit = {
    factoryMap(typeOf[T]) = factory
  }

  registerFactory(() => Validators.forArray[Double](ordered = true))
  registerFactory(() => Validators.forArray2d[Double](dim1Ordered = true, dim2Ordered = true))
  registerFactory(() => Validators.forList[Double](ordered = true))
  registerFactory(() => Validators.forList2d[Double](dim1Ordered = true, dim2Ordered = true))

  def create[T]()(implicit tt: TypeTag[T]): (T, T) => Boolean = {
    create(typeOf[T])
  }

  def create[T](elemType: Type): (T, T) => Boolean = {
    factoryMap.get(elemType) match {
      case Some(fact) => fact().asInstanceOf[(T,T)=>Boolean]
      case None => (a, b) => FeatureFactory.create[T](elemType).isEqual(a, b)
    }
  }

}
